daml-foundations/daml-ghc/daml-stdlib-src/DA/Internal/Template.daml

```diff
@@ -89,12 +89,30 @@ class Template c where
     internalArchive = magic @"archive"
 
 
+-- Deliberately not exported.
+data Consuming = PreConsuming  -- Archive before executing exercise body.
+               | PostConsuming -- Execute exercise body then archive.
+               | NonConsuming  -- Don't archive.
+
+-- | HIDE
+nonconsuming : NoEvent c e -> Consuming
+nonconsuming _ = NonConsuming
+
+-- | HIDE
+preconsuming : NoEvent c e -> Consuming
+preconsuming _ = PreConsuming
+
+-- | HIDE
+postconsuming : NoEvent c e -> Consuming
+postconsuming _ = PostConsuming
+
 class Template c => Choice c e r | c e -> r where
     -- | HIDE
-    consuming : NoEvent c e -> ChoiceType
-    -- must be a lambda since otherwise the default-dictionary method causes a runtime loop as the method
-    -- exists in the dictionary, takes the dictionary, and can't be eta expanded.
-    consuming _ = Consuming
+    consuming : NoEvent c e -> Consuming
+    -- Must be a lambda since otherwise the default-dictionary method
```

Comment:
What does this comment actually express? How could this not be a lambda?

Reply:
It could evaluate to a lambda, rather than being a lambda.

Reply:
Oh I see, it had to be a _syntactic_ lambda. Anyway, that's no longer necessary since I re-implemented the dictionary sanitization. I've prepared a PR to remove the comment: https://github.com/digital-asset/daml/pull/628

Reply:
The comment was pre-existing. I also had doubts about it, removed it, changed the implementation to `consuming _ = preconsuming` and found it did indeed cause non-termination so was forced to reinstate it.

Reply:
When I do `consuming = preconsuming`, the test suite still passes. I'll amend the other PR to use that implementation.